core/src/main/java/org/elasticsearch/search/aggregations/support/InternalOrder.java

```diff
@@ -0,0 +1,603 @@
+     *
+     * @param order    bucket ordering strategy to check. If this is a {@link CompoundOrder} the first element will be
+     *                 check instead.
+     * @param expected expected  bucket ordering strategy.
+     * @return {@code true} if the order matches, {@code false} otherwise.
+     */
+    private static boolean isOrder(BucketOrder order, BucketOrder expected) {
+        if (order == expected) {
+            return true;
+        } else if (order instanceof CompoundOrder) {
+            // check if its a compound order with the first element that matches
+            List<BucketOrder> orders = ((CompoundOrder) order).orderElements;
+            if (orders.size() >= 1) {
+                return isOrder(orders.get(0), expected);
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Contains logic for reading/writing {@link BucketOrder} from/to streams.
+     */
+    public static class Streams {
+
+        /**
+         * Read a {@link BucketOrder} from a {@link StreamInput}.
+         *
+         * @param in stream with order data to read.
+         * @return order read from the stream
+         * @throws IOException on error reading from the stream.
+         */
+        public static BucketOrder readOrder(StreamInput in) throws IOException {
+            byte id = in.readByte();
+            switch (id) {
+                case COUNT_DESC_ID:
+                    return InternalOrder.COUNT_DESC;
+                case COUNT_ASC_ID:
+                    return InternalOrder.COUNT_ASC;
+                case KEY_DESC_ID:
+                    return InternalOrder.KEY_DESC;
+                case KEY_ASC_ID:
+                    return InternalOrder.KEY_ASC;
+                case Aggregation.ID:
+                    boolean asc = in.readBoolean();
+                    String key = in.readString();
+                    return new InternalOrder.Aggregation(key, asc);
+                case CompoundOrder.ID:
+                    int size = in.readVInt();
+                    List<BucketOrder> compoundOrder = new ArrayList<>(size);
+                    for (int i = 0; i < size; i++) {
+                        compoundOrder.add(Streams.readOrder(in));
+                    }
+                    return new CompoundOrder(compoundOrder, false);
+                default:
+                    throw new RuntimeException("unknown order id [" + id + "]");
+            }
+        }
+
+        /**
+         * ONLY FOR HISTOGRAM ORDER: Backwards compatibility logic to read a {@link BucketOrder} from a {@link StreamInput}.
+         *
+         * @param in stream with order data to read.
+         * @return order read from the stream
+         * @throws IOException on error reading from the stream.
+         */
+        public static BucketOrder readHistogramOrder(StreamInput in) throws IOException {
+            if (in.getVersion().onOrAfter(Version.V_6_0_0_alpha1_UNRELEASED)) {
```

Comment:
I think you should be able to do this in a unit test for this class when you create the `StreamInput` you should be able to set the version